### DISCUSSION TITLE
`configureVitest` plugin hook

### DISCUSSION THREAD

**Discussion by sheremet-va** [vitest-dev/vitest#7104] [Feedback]
The teams wants to propose a Vitest-specific [plugin hook](https://vite.dev/guide/using-plugins.html) called `configureVitest` (similar to `configureServer`). The hook will be called after the workspace projects are created and filtered and before the reporter resolution.

```ts
import type { Plugin } from 'vitest/node'
import type { Plugin } from 'vite' // augmented

interface VitestPluginContext {
  project: TestProject; // the project to which the server belongs to
  addTestProject(config: string | InlineConfig): Promise<TestProject>
  addBrowserCommand(name: string, command: BrowserCommand): void
}

const plugin: Plugin = {
  name: 'vitest-hook',
  async configureVitest({ project, addTestProject }: VitestPluginContext) {
	// ...

    project.globalConfig.reporters.push(['@storybook/vitest-reporter', {}])
    project.globalConfig.coverage.enabled = false
    // this is injected into vitest.projects. The name validation is still enabled
    const project = await addTestProject('/my-custom-config.js')
    // the difference from vitest.projects.push() is that plugin authors 
    // don't need to create the project themselves
  }
}
```

This hook is mostly useful to library authors or in very advanced use cases where you need to access the [public Node.js API](https://main.vitest.dev/advanced/api/).

Vitest already provides [reporter API](https://main.vitest.dev/advanced/reporters.html) that covers the test runner lifecycle hooks and some additional hooks like `onWatcherStart` or `onPathsCollected`, we will not change that and keep adding new hooks to reporters if the need arises.

However while you can already use Vite plugins to inject some config values in the `config` hook, it is not possible to manipulate the root Vitest config from the workspace projects:

```ts
export default defineConfig({
  test: {
    workspace: [
      {
        plugins: [
          {
            config() {
              // cannot inject reporter or change coverage or any root option
            }
          },
        ],
      },
    ],
  },
})
```

With `configureVitest` hook you can directly modify the resolved global Vitest config. It is still recommended to use the `config` plugin hook to inject project-level options like `setupFiles` or `maxConcurrency`.

This plugin hook will not be called for server created to serve browser assets in the [browser mode](https://main.vitest.dev/guide/browser/). The hook will be called in parallel for all plugins in a single project. Project plugins will be called sequentially:

```ts
for (const project of vitest.projects) {
  await Promise.all(project.plugins.map(p => p.configureVitest(context))
}
```

Note that it's possible to have the same config in separate workspace projects, so it's recommended to validate the config first before injecting or changing values.

**Comment by JReinhold** [comment]
> The hook will be called in parallel for all plugins in a single project. Project plugins will be called sequentially

How does this work with `addTestProject()`? What if I use `addTestProject()` to add a project, which itself has a plugin with the `configureVitest` hook? Will that be supported, or is this more similar to the plugins limitation of the `config` hook in Vitest, where you can modify the config from a plugin, except you can't modify the `plugins` property?

**Reply by sheremet-va** [reply]
I don't think they should chain, personally, but it's an interesting question! I wonder what others think. This situation looks similar to what Vite has with `configResolved` - if you push plugins, they won't call `config` hook

**Reply by JReinhold** [reply]
I agree, I think it's a fair limitation.

**Comment by JReinhold** [comment]
It sounds to me like `addTestProject` is a way to dynamically add new projects. Will that be supported with the `--project=my-project` CLI flag? Ie. can I still use that flag to filter for a specific project, even if that project only exists because it's dynamically added with `addTestProject()`?

**Reply by sheremet-va** [reply]
Currently proposal says this:

> The hook will be called after the workspace projects are created and filtered and before the reporter resolution

So no, if there is no project already, it won't work. But it's a good topic to discuss how this could work 🤔

**Reply by JReinhold** [reply]
Sorry I missed that. I _think_ that could be a dealbreaker for many, high risk it would make Storybook not use `addTestProject` at least, if it meant users couldn't filter on the project.

**Comment by JReinhold** [comment]
This is a nit, and there might be context I'm missing, but I find it slightly weird that `globalConfig` is a property of the `project`, instead of just being a sibling to it.

**Reply by sheremet-va** [reply]
project is a [`TestProject`](https://main.vitest.dev/advanced/api/test-project.html) API, it's already established. You can get the same config via `project.vitest.config`

**Comment by JReinhold** [comment]
What is special about browser commands, that they warrant a specific `addBrowserCommand` function here? I don't personally know what the use case is, so I'm just wondering why a function for _this_ and not any of all the other configurations?

**Reply by sheremet-va** [reply]
Personally, I am just not satisfied with the current API - too much nesting: https://main.vitest.dev/guide/browser/commands.html#custom-commands

```ts
export default function BrowserCommands(): Plugin {
  return {
    name: 'vitest:custom-commands',
    config() {
      return {
        test: {
          browser: {
            commands: {
              myCustomCommand() {
                // do something
              },
            }
          }
        }
      }
    }
  }
}
```

**Reply by JReinhold** [reply]
Ah I see. I personally don't see that as an issue, but we don't have to agree on that.

However your argument could be used to say you should also have an `addBrowserOrchestratorScripts()`, `setCoverageThresholds()` etc.

**Reply by sheremet-va** [reply]
scripts and thresholds are just strings and objects. Commands can have a lot of nested logic

**Comment by JReinhold** [comment]
> `addTestProject(config: string | InlineConfig)`

My recommendation would be to stay away from resolvable strings as configs, and require explicit configurations instead. To me, one of Vite's core strengths (and Storybook's weakness) is the fact that you have to import plugins, you can't just reference them with a string like `"my-package/plugin"` and then Vite will try to resolve that.

We have _so_ many issues with the string-based approach in Storybook - what if the dependency is hoisted in a monorepo? what about Yarn PnP? What is the relative path relative _to_? etc.

If this is indeed just a shortcut and doesn't provide any extra value over just `InlineConfig`, I'd recommend making the plugin author import the project config themselves, using their own strategy, to minimize complexity and use cases.

Another argument, is that it would be hard/impossible to merge configs using the resolvable string. I can imagine the Storybook plugin would use `addTestProject()` internally, so we didn't require users to create that file themselves. Then we'd make our `storybookTest()` plugin function take in a partial config with specific project-config overrides they'd want, and we'd merge those overrrides before passing them in to `addTestProject({ ...mergedConfig })`. We wouldn't be able to do that merging if we only passed in `addTestProject('@storybook/addon-test/internal/vitest-project')`.

**Reply by sheremet-va** [reply]
> I'd recommend making the plugin author import the project config themselves, using their own strategy, to minimize complexity and use cases.

I don't see how this minimizes complexity? Vitest and Vite already import configs with `resolveConfig` method. This is the same as calling `vite.createServer({ configFile: path })` (this is how workspace projects already work).

> Then we'd make our storybookTest() plugin function take in a partial config with specific project-config overrides they'd want, and we'd merge those overrrides before passing them in to `addTestProject({ ...mergedConfig })`. We wouldn't be able to do that merging if we only passed in `addTestProject('@storybook/addon-test/internal/vitest-project')`.

Not really, you would just use `configFile` property on the `InlineConfig`:

```ts
configureVitest({ project }) {
  await addTestProject({
    configFile: project.vite.config.configFile, // reuse the same config file
    // ... custom override properties
  })
}
```

**Reply by JReinhold** [reply]
ah I see, I didn't know that.

**Comment by sheremet-va** [comment]
I implemented a basic idea in https://github.com/vitest-dev/vitest/pull/7349

Some notable changes:
- `addTestProject` was renamed to `injectTestProjects` and now returns an array of projects
- projects with the name that was already defined will throw an error
- `injectTestProjects` will not inject a project if it was filtered out unless you specify `force: true` when injecting it
- plugin context now has direct access to `vitest` property instead of accessing it via `project.vitest`
- `addBrowserCommand` is postponed for now

The current context looks like this:

```ts
export interface VitestPluginContext {
  vitest: Vitest
  project: TestProject
  injectTestProjects: (config: TestProjectConfiguration | TestProjectConfiguration[]) => Promise<TestProject[]>
}
```

Documentation: https://deploy-preview-7349--vitest-dev.netlify.app/advanced/api/plugin.html
